contrib/babelfishpg_tsql/runtime/functions.c

```diff
@@ -1916,6 +1920,186 @@ checksum(PG_FUNCTION_ARGS)
+ *	This function performs a search to find the partition number
+ *	for a given value in a specified partition function by retrieving
+ *	the partition function metadata, and performing a binary search 
+ *	on the sorted array of partition range values.
+ *
+ *	Returns the index of the partition to which the input value belongs.
+ */
+Datum
+search_partition(PG_FUNCTION_ARGS)
+{
+	char			*partition_func_name = text_to_cstring(PG_GETARG_TEXT_P(0));
+	int32			result;
+	Relation		rel;
+	HeapTuple		tuple;
+	SysScanDesc		scan;
+	ScanKeyData		scanKey[2];
+	int16			dbid;
+	ArrayType		*values;
+	bool			isnull;
+	Datum			arg;
+	Oid			argtypeid;
+	char			*func_param_typname = NULL;
+	Oid			func_param_typoid;
+	Oid			sqlvariant_typoid;
+	Datum			*range_values;
+	bool			*nulls;
+	int			nelems;
+	Oid			basetype_oid;
+	Oid			opclass_oid;
+	Oid			opfamily_oid;
+	Oid			cmpfunction_oid;
+	tsql_compare_context	cxt;
+
+	if (!PG_ARGISNULL(2)) /* Database is specified. */
+	{
+		char *db_name = text_to_cstring(PG_GETARG_TEXT_P(2));
+		dbid = get_db_id(db_name);
+		if (!DbidIsValid(dbid))
+			ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					errmsg("Invalid database name '%s'.", db_name)));
+		pfree(db_name);
+	}
+	else /* Database is not specified. */
+		dbid = get_cur_db_id();
+	
+	/* Get OID of sql_variant type. */
+	sqlvariant_typoid = (*common_utility_plugin_ptr->get_tsql_datatype_oid) ("sql_variant");
+	
+	/*
+	 * Get metadata of partition function for the provided partition
+	 * function name, if it exists in the provided database.
+	 */
+	rel = table_open(get_bbf_partition_function_oid(), AccessShareLock);
+	
+	ScanKeyInit(&scanKey[0],
+				Anum_bbf_partition_function_dbid,
+				BTEqualStrategyNumber, F_INT2EQ,
+				Int16GetDatum(dbid));
+
+	ScanKeyEntryInitialize(&scanKey[1], 0,
+				Anum_bbf_partition_function_name,
+				BTEqualStrategyNumber, InvalidOid,
+				tsql_get_server_collation_oid_internal(false),
+				F_TEXTEQ, CStringGetTextDatum(partition_func_name));
+	
+	/* Scan using index. */
+	scan = systable_beginscan(rel, get_bbf_partition_function_pk_idx_oid(),
+			false, NULL, 2, scanKey);
+	
+	tuple = systable_getnext(scan);
+	if (HeapTupleIsValid(tuple))
+	{
+		func_param_typname = TextDatumGetCString(heap_getattr(tuple, Anum_bbf_partition_function_input_parameter_type, RelationGetDescr(rel), &isnull));
+		values = DatumGetArrayTypeP(heap_getattr(tuple, Anum_bbf_partition_function_range_values, RelationGetDescr(rel), &isnull));
+		deconstruct_array(values, sqlvariant_typoid,
+					-1, false, 'i', &range_values, &nulls, &nelems);
+	}
+	systable_endscan(scan);
+	table_close(rel, AccessShareLock);
+
+	/* Raise error if provided partition function doesn't exist in the provided database. */
+	if (!func_param_typname)
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					errmsg("Invalid object name '%s'.", partition_func_name)));
+	
+	/* NULL values will always belong to first partition. */
+	if (PG_ARGISNULL(1))
```

Comment:
I see, need to add comment about the chronology of the error/behaviour in TSQL at the top of the function

Reply:
addressed.